iommi/part.py

```diff
@@ -65,15 +67,21 @@ class Part(Traversable):
     extra: Dict[str, Any] = Refinable()
     extra_evaluated: Dict[str, Any] = Refinable()  # not EvaluatedRefinable because this is an evaluated container so is special
     endpoints: Namespace = Refinable()
+    assets: Namespace = Refinable()
 
     @reinvokable
     @dispatch(
         extra=EMPTY,
         include=True,
+        assets=EMPTY,
     )
-    def __init__(self, *, endpoints: Dict[str, Any] = None, include, **kwargs):
+    def __init__(self, *, endpoints: Dict[str, Any] = None, include, assets=None, **kwargs):
         super(Part, self).__init__(include=include, **kwargs)
         collect_members(self, name='endpoints', items=endpoints, cls=Endpoint)
+        # What I would love to do here is collect all my childrens assets, but I think
+        # I can't do that before my children are properly bound.  So instead I store
+        # only my own assets in assets and later do a grand collecting exercise
```

Comment:
I think you could treat the assets much like we do with the endpoints.
- Use collect members here. (And have some `Asset` class to just hold the value of the asset, just like `Endpoint`)
- Use `bind_members(..)` in `on_bind` and have the `Asset` have an `on_bind` that does something in the lines of `setattr_path(parent.iommi_root().assets, **{self._name, self.asset_value})`.

Then you could in your style do things like:
```
    Field.shortcuts.choice.assets.select2_js.include=False,
    Field.choice.assets.my_fancy_select=...
```

Does this make sense?

Reply:
I didn't understand this, so I gave Johan a call :P

The idea being here that parts have their assets, but in on_bind they will write them to the root part, which is the one we then render.

Reply:
I like Johan's idea.  It's both more consistent with the rest of iommi and the direct writing to the root is also cheaper than the walking of the tree desparately looking for an asset.

Would this also require some changes inside style to make the new Asset class work, or will that just work (I'm thinking of line 93 in style.py).

Reply:
I think currently the assets of the style comes a different way, but I think they could be merged in `render_root` in `Part` somehow. (i.e. we merge the `style_assets` with the assets that has been put on the root by child components during bind.)

Reply:
A random remark.  I think I have written a large portion of what is necessary, but currently it's not working.  And while hunting the bug I had to read `collect_members`, that function could really do with a doc string...

Reply:
The whole member collect/bind/reinvoke/laziness machinery is a bit hairy. I think it works fairly well, but the implementation is far from obvious.

Reply:
And another thing I can't get my head around.  Where is Endpoint.name actually used?  I see that the iommi_path is obviously used (and thereby Endpoint._name) but I don't see any code that makes use of Endpoint's name field.  Is there some magic that makes name and _name work together?  Or is Endpoint.name some sad leftover...

Reply:
Just a leftover from before it became a `Traversable`.

Reply:
Another thing would you be surprised if I tell you that the first node that render_root is called on is not the node you get when you call iommi_root inside on_bind ?  I very much was but it was the result of a long debugging session yesterday (and I don't yet know why but I'm sure I will eventually).

Reply:
I added docstrings for `collect_members` and `bind_members`, and moved them up in the file as they are the most important parts of that module.

Reply:
Thanks all.  So here is the weird thing that currently has me stumped:

I added a print statement at the top of Traversable.bind:

```python
    def bind(self, *, parent=None, request=None):
        print("bind called on", id(self), "name", self._name, "parent",
              (id(parent), parent._name) if parent is not None else "None")
```

and one on the top of render_root in parts.py:

```python
    print("render_root called on", id(part), "name", part._name, "parent",
          (id(part._parent), part._parent._name) if part._parent is not None else "None",
          "iommi_root", id(part.iommi_root()))
```
 
And then I render the two table example and the beginning of the output looks like this:


```text
bind called on 4378140120 name None parent None
bind called on 4378771528 name parts parent (4378709856, 'root')
bind called on 4378771752 name endpoints parent (4378709856, 'root')
bind called on 4378771920 name assets parent (4378709856, 'root')
render_root called on 4378709856 name root parent None iommi_root 4378709856
```

So why is the first node that bind is called upon not the same node as the node that is passed to render_root.  In fact how is 4378709856 ever bound? It never appears in the log as a argument of bind (only as a parent of a node passed to bind)...

Is that expected or some subtle bug that never mattered so far? 

Thoughts?

Reply:
`bind` returns a new object, so I would assume the first line (id 4378140120) results in bind returning the object you're looking for (id 4378709856).

Reply:
Damn.  Okay very true.  In which case I was barking up the wrong tree.  Thanks

Reply:
Oh but I think I understand now.  @jlubcke is it possible that your design doesn't work with the current structure `render_root` because only `part.__html__` actually causes all the binds of all the children to trigger?  Or with other words I'm currently trying to read the assets of the parts at the top render_root but at that time the lazyness in bind will not have actually caused them to be evaluated and done their side effects...  (Which is the actual problem I was observing that the assets always stayed empty).  

If you agree I would move `content=part.__html__(**render`) to the top of `render_root`

Reply:
Good point. All the lazy binds need to have happened once we get around to render the assets. So moving the `__html__` should do the trick.